src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,96 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
     res
 end
+
+
+Base.transpose(::AbstractDataFrame, args...; kwargs...) =
+    MethodError("`transpose` not defined for `AbstractDataFrame`s. Try `permutedims` instead")
+    # ↑ is 94 char, is that OK? (there are a couple other lines in this file that go over)
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                dest_namescol::Union{Symbol,Int};
+                makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex; makeunique::Bool=false)
+    permutedims(df::AbstractDataFrame; makeunique::Bool=false)
+
+Turn a `DataFrame` on its side such that rows become columns
+and the column indexed by `src_namescol` becomes a header.
+In the resulting `DataFrame`,
+The header of `df` will become the first column
+with name specified by `dest_namescol`
+
+# Arguments
+- `df` : the `AbstractDataFrame`
+- `src_namescol` : the column that will become the new header.
+  Defaults to first column.
+- `dest_namescol` : the name of the first column in the returned `DataFrame`.
+  Defaults to the same name as `src_namescol`.
+- `makeunique` : if `false` (the default), an error will be raised
+  if duplicate names are found; if `true`, duplicate names will be suffixed
+  with `_i` (`i` starting at 1 for the first duplicate).
+
+Note: The eltypes of columns in resulting `DataFrame` will depend
+on the eltypes of _all_ input columns.
+That is, if the source `DataFrame` contains `String` and `Int` columns,
+all resulting columns will have eltype `Any`.
+If the source has a mix of numeric types (eg. `Float64` and `Int`),
+all columns in resulting `DataFrame` will be promoted to `Float64`.
+
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=["x", "y"], b=rand(2), c=[1,2], d=rand(Bool,2));
+
+julia> df2 = DataFrame(a=["x", "y"], b=[1, "str"], c=[1,2], d=rand(Bool,2));
+
+julia> permutedims(df1) # note the column type
+3×3 DataFrame
+│ Row │ a      │ x        │ y        │
+│     │ String │ Float64  │ Float64  │
+├─────┼────────┼──────────┼──────────┤
+│ 1   │ b      │ 0.982197 │ 0.263357 │
+│ 2   │ c      │ 1.0      │ 2.0      │
+│ 3   │ d      │ 0.0      │ 1.0      │
+
+julia> permutedims(df2)
+3×3 DataFrame
+│ Row │ a      │ x   │ y   │
+│     │ String │ Any │ Any │
+├─────┼────────┼─────┼─────┤
+│ 1   │ b      │ 1   │ str │
+│ 2   │ c      │ 1   │ 2   │
+│ 3   │ d      │ 0   │ 0   │
+````
+"""
+function Base.permutedims(df::AbstractDataFrame, src_namescol::ColumnIndex,
+                          dest_namescol::Union{Symbol, AbstractString};
+                          makeunique::Bool=false)
+
+    nrow(df) > 0 || throw(ArgumentError("`permutedims` not defined for data frame with 0 rows"))
+    eltype(df[!, src_namescol]) <: SymbolOrString || throw(
+            ArgumentError("src_namescol must have eltype `Symbol` or `AbstractString`"))
+
+    df_notsrc = df[!, Not(src_namescol)]
+    df_permuted = DataFrame([names(df_notsrc)], [dest_namescol])
```

Comment:
this seems cleaner:
```suggestion
    df_permuted = DataFrame(dest_namescol => names(df_notsrc))
```

Reply:
also a decision to be made is if we want to produce `String`s or `Symbol`s here (see a similar situation in `stack`/`unstack` - maybe we should have a kwarg allowing to decide it - and then what should be the default?).

Reply:
Will leave this open waiting for @nalimilan to respond. My take is that since `names` went to strings, we should use strings. I'm happy to add a keyword though if there's consensus (should maybe allow `Int` as a possibility too in that case?)

Reply:
Sure. In `stack` we default to `String`, so for now we could leave it as is and add a kwarg later if needed (similarly to promotion).

Reply:
Agreed. Contrary to `stack`, each name appears only once, so there's little point in using a `PooledArray` or a `CategoricalArray`.

Reply:
But the point is not about being pooled, but if the column should be of eltype: `String` or `Symbol` (in `stack` we allow the user to choose one of them).

Reply:
Yes, I know, I was commenting on the fact that a keyword argument is less needed because `CategoricalValue` is less useful. Indeed we could add an argument later to get symbols instead of strings, but that doesn't sound too important.

Reply:
:+1: OK, will leave as-is for now.